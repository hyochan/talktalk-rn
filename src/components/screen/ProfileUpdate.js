// @flow
import React, { Component } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  ScrollView,
} from 'react-native';
import Icon5, { FA5Style } from 'react-native-vector-icons/FontAwesome5';

import styled from 'styled-components/native';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import Button from '../shared/Button';
import TextInput from '../shared/TextInput';
import { IC_MASK } from '../../utils/Icons';

import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

type Styles = {
  container: ViewStyle,
  scrollView: ViewStyle,
  scrollViewContainer: ViewStyle,
  wrapper: ViewStyle,
  btnWrapper: ViewStyle,
  btnLogout: ViewStyle,
  txtLogout: TextStyle,
  btnUpdate: ViewStyle,
  txtUpdate: TextStyle,
};

const styles: Styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    flexDirection: 'column',
    alignItems: 'center',
  },
  scrollView: {
    alignSelf: 'stretch',
  },
  scrollViewContainer: {
    justifyContent: 'center',
    alignItems: 'center',
  },
  wrapper: {
    marginTop: 48,
    width: '78%',

    flexDirection: 'column',
    alignItems: 'center',
  },
  btnWrapper: {
    width: '100%',
    justifyContent: 'space-between',
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 24,
    marginBottom: 48,
  },
  btnLogout: {
    backgroundColor: 'transparent',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 1,
    width: 136,
    height: 60,
    borderColor: colors.dodgerBlue,
    marginRight: 4,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtLogout: {
    fontSize: 16,
    fontWeight: 'bold',
    color: colors.dodgerBlue,
  },
  btnUpdate: {
    backgroundColor: colors.dodgerBlue,
    borderColor: colors.dodgerBlue,
    borderRadius: 4,
    borderWidth: 1,
    width: 136,
    height: 60,
    marginLeft: 4,
    shadowColor: colors.dodgerBlue,
    shadowOffset: {
      width: 0,
      height: 10,
    },
    shadowRadius: 4,
    shadowOpacity: 0.3,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtUpdate: {
    fontSize: 16,
    fontWeight: 'bold',
    color: 'white',
  },
});

type Props = {
  navigation: any,
};
type State = {
  isUpdating: boolean;
  displayName: string;
  statusMsg: string;
  photoURL: string;
};

class Screen extends Component<Props, State> {
  static navigationOptions = {
    title: getString('MY_PROFILE'),
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      isUpdating: false,
      displayName: '',
      statusMsg: '',
      photoURL: '',
    };
  }

  render() {
    return (
      <View style={styles.container}>
        <ScrollView
          style={styles.scrollView}
          contentContainerStyle={styles.scrollViewContainer}
        >
          <View style={styles.wrapper}>
            <TouchableOpacity
              activeOpacity={0.5}
              onPress={this.onPressImg}
            >
              <Icon5 name="meh" size={80} color={colors.dusk} light/>
            </TouchableOpacity>
            <TextInput
              style={{ marginTop: 80 }}
              txtLabel={getString('NAME')}
              txtHint={ getString('NAME') }
              txt={ this.state.displayName }
              onTextChanged={ (text) => this.onTextChanged('DISPLAY_NAME', text)}
            />
            <TextInput
              style={{ marginTop: 24 }}
              txtLabel={getString('STATUS_MSG')}
              txtHint={ getString('STATUS_MSG') }
              txt={ this.state.statusMsg }
              onTextChanged={ (text) => this.onTextChanged('STATUS_MSG', text)}
            />
            <View style={styles.btnWrapper}>
              <Button
                id='logout'
                onPress={() => this.onLogout()}
                isWhite
              >{getString('LOGOUT')}</Button>
              <View style={{ width: 8 }}/>
              <Button
                id='update'
                isLoading={this.state.isUpdating}
                onPress={() => this.onUpdate()}
              >{getString('UPDATE')}</Button>
            </View>
          </View>
        </ScrollView>
      </View>
    );
  }

  onLogout = () => {
    this.props.navigation.navigate('AuthStackNavigator');
  }

  onUpdate = () => {
    this.setState({ isUpdating: true }, () => {
      try {
        this.props.navigation.goBack();
      } catch (err) {
        console.error(err);
      } finally {
        this.setState({ isUpdating: false });
      }
    });
  }

  onTextChanged = (type: string, text: string) => {
    switch (type) {
      case 'DISPLAY_NAME':
        this.setState({ displayName: text });
        break;
      case 'STATUS_MSG':
        this.setState({ statusMsg: text });
        break;
    }
  }

  onPressImg = () => {
  }
}

export default Screen;
