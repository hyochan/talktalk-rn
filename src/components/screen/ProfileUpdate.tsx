// @flow
import React, {
  Component,
  useState,
  useEffect,
} from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  ScrollView,
} from 'react-native';
import { NavigationScreenProp, NavigationStateRoute } from 'react-navigation';
import Icon5, { FA5Style } from 'react-native-vector-icons/FontAwesome5';

import styled from 'styled-components/native';

import Button from '../shared/Button';
import TextInput from '../shared/TextInput';
import { IC_MASK } from '../../utils/Icons';

import { ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

const StyledContainer = styled.View`
  flex: 1;
  background-color: white;
  flex-direction: column;
  align-items: center;
`;

const StyledScrollView = styled.ScrollView`
  align-self: stretch;
`;

const StyledWrapper = styled.View`
  margin-top: 48;
  width: 78%;

  flex-direction: column;
  align-items: center;
`;

const StyledBtnWrapper = styled.View`
  width: 100%;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  margin-top: 24;
  margin-bottom: 48;
`;

interface IProps {
  navigation: NavigationScreenProp<NavigationStateRoute<any>>;
}

interface IState {
  isUpdating: boolean;
  displayName: string;
  statusMsg: string;
  photoURL: string;
}

function Screen(props: IProps) {
  const [isUpdating, setIsUpdating] = useState(false);
  const [displayName, setDisplayName] = useState('');
  const [statusMsg, setStatusMsg] = useState('');
  const [photoURL, setPhotoURL] = useState('');

  useEffect(() => {
    if (isUpdating) {
      try {
        props.navigation.goBack();
      } catch (err) {
        console.error(err);
      } finally {
        setIsUpdating(false);
      }
    }
  }, [isUpdating]);

  const onLogout = () => {
    props.navigation.navigate('AuthStackNavigator');
  };

  const onUpdate = () => {
    setIsUpdating(true);
  };

  const onTextChanged = (type: string, text: string) => {
    switch (type) {
      case 'DISPLAY_NAME':
        setDisplayName(text);
        break;
      case 'STATUS_MSG':
        setStatusMsg(text);
        break;
    }
  };

  const onPressImg = () => {
  };

  return (
    <StyledContainer>
      <StyledScrollView
        contentContainerStyle={{
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        <StyledWrapper>
          <TouchableOpacity
            activeOpacity={0.5}
            onPress={onPressImg}
          >
            <Icon5 name='meh' size={80} color={colors.dusk} light/>
          </TouchableOpacity>
          <TextInput
            testID='input_name'
            style={{ marginTop: 80 }}
            txtLabel={getString('NAME')}
            txtHint={ getString('NAME') }
            txt={ displayName }
            onTextChanged={ (text) => onTextChanged('DISPLAY_NAME', text)}
          />
          <TextInput
            testID='input_status'
            style={{ marginTop: 24 }}
            txtLabel={getString('STATUS_MSG')}
            txtHint={ getString('STATUS_MSG') }
            txt={ statusMsg }
            onTextChanged={ (text) => onTextChanged('STATUS_MSG', text)}
          />
          <StyledBtnWrapper>
            <Button
              testID='logout_btn'
              onPress={() => onLogout()}
              isWhite
            >{getString('LOGOUT')}</Button>
            <View style={{ width: 8 }}/>
            <Button
              testID='update_btn'
              isLoading={isUpdating}
              onPress={() => onUpdate()}
            >{getString('UPDATE')}</Button>
          </StyledBtnWrapper>
        </StyledWrapper>
      </StyledScrollView>
    </StyledContainer>
  );
}

export default Screen;
