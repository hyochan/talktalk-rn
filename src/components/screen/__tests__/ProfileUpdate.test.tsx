import 'react-native';
import * as React from 'react';
import ProfileUpdate from '../ProfileUpdate';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

describe('rendering test', () => {
  const props = {
    navigation: {
      navigate: jest.fn(),
    },
  };
  const component: React.ReactElement = <ProfileUpdate {...props} />;

  it('renders as expected', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
  });
});

describe('interaction', () => {
  let rendered: renderer.ReactTestRenderer;
  let root: renderer.ReactTestInstance;
  let testingLib: any;
  let props: any;
  let component: React.ReactElement;

  beforeAll(() => {
    rendered = renderer.create(component);
    root = rendered.root;
  });

  beforeEach(() => {
    props = {
      navigation: {
        navigate: jest.fn(),
        goBack: jest.fn(),
      },
    };
    component = <ProfileUpdate {...props} />;
    testingLib = render(component);
  });

  it('should fireEvent when update button pressed', () => {
    fireEvent(testingLib.getByTestId('update_btn'), 'press');
    // expect(props.navigation.goBack).toHaveBeenCalledTimes(1);
  });

  it('should fireEvent when logout button pressed', () => {
    fireEvent(testingLib.getByTestId('logout_btn'), 'press');
    expect(props.navigation.navigate).toHaveBeenCalledWith('AuthStackNavigator');
  });

  it('should changeText when display name changed', () => {
    const inputName = testingLib.getByTestId('input_name');
    inputName.props.onTextChanged('name');
    expect(inputName.props.txt).toEqual('name');
  });

  it('should changeText when status message changed', () => {
    const inputStatus = testingLib.getByTestId('input_status');
    inputStatus.props.onTextChanged('status');
    expect(inputStatus.props.txt).toEqual('status');
  });
});
