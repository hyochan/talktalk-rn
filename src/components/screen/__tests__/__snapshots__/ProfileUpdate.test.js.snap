// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders as expected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <ScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
    style={
      Object {
        "alignSelf": "stretch",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "marginTop": 48,
          "width": "78%",
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.5}
        onPress={[Function]}
      >
        <FA5iconClass
          color="rgb(65,77,107)"
          light={true}
          name="meh"
          size={80}
          solid={false}
        />
      </TouchableOpacity>
      <Shared
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 80,
          }
        }
        txt=""
        txtHint="name"
        txtLabel="name"
      />
      <Shared
        labelStyle={
          Object {
            "color": "rgb(134,154,183)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        labelStyleFocus={
          Object {
            "color": "rgb(58,139,255)",
            "fontSize": 12,
            "marginBottom": 8,
          }
        }
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 24,
          }
        }
        txt=""
        txtHint="Status message"
        txtLabel="Status message"
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginBottom": 48,
            "marginTop": 24,
            "width": "100%",
          }
        }
      >
        <Button
          activeOpacity={0.5}
          height={60}
          id="logout"
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          isWhite={true}
          onPress={[Function]}
          width={136}
        >
          Logout
        </Button>
        <View
          style={
            Object {
              "width": 8,
            }
          }
        />
        <Button
          activeOpacity={0.5}
          height={60}
          id="update"
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          isWhite={false}
          onPress={[Function]}
          width={136}
        >
          Update
        </Button>
      </View>
    </View>
  </ScrollView>
</View>
`;
